calico_containers/tests/unit/calicoctl_test.py

```diff
@@ -216,6 +221,211 @@ def test_validate_arguments(self, case, sys_exit_called):
+                     '/usr/share/docker/plugins']
+        )
+        m_docker_client.start.assert_called_once_with(container)
+        m_attach_and_stream.assert_called_once_with(container)
+
+
+    @patch('sys.exit', autospec=True)
+    @patch('calico_ctl.node.os', autospec=True)
+    @patch('calico_ctl.node.check_system', autospec=True)
+    @patch('calico_ctl.node.get_host_ips', autospec=True)
+    @patch('calico_ctl.node.warn_if_unknown_ip', autospec=True)
+    @patch('calico_ctl.node.warn_if_hostname_conflict', autospec=True)
+    @patch('calico_ctl.node.install_kubernetes', autospec=True)
+    def test_node_start_oserror(self, m_install_kube,
+                        m_warn_if_hostname_conflict, m_warn_if_unknown_ip,
+                        m_get_host_ips, m_check_system, m_os, m_sys_exit):
+        """
+        Test node_start function in calicoctl node command
+        """
+        # Set up mock objects
+        m_os.path.exists.return_value = False
+        m_get_host_ips.return_value = []
+        m_install_kube.side_effect = OSError
+
+        # Set up arguments
+        node_image = 'node_image'
+        log_dir = './log_dir'
+        ip = ''
+        ip6 = 'aa:bb::zz'
+        as_num = ''
+        detach = False
+        kubernetes = True
+
+        # Test expecting OSError exception
+        self.assertRaises(OSError, node.node_start,
+                          node_image, log_dir, ip, ip6, as_num, detach, kubernetes)
+
+    @patch('sys.exit', autospec=True)
+    @patch('calico_ctl.node.os', autospec=True)
+    @patch('calico_ctl.node.check_system', autospec=True)
+    @patch('calico_ctl.node.get_host_ips', autospec=True)
+    @patch('calico_ctl.node.warn_if_unknown_ip', autospec=True)
+    @patch('calico_ctl.node.warn_if_hostname_conflict', autospec=True)
+    @patch('calico_ctl.node.install_kubernetes', autospec=True)
+    @patch('calico_ctl.node.client', autospec=True)
+    @patch('calico_ctl.node.docker_client', autospec=True)
+    def test_node_start_apierror(self, m_docker_client, m_client, m_install_kube,
+                        m_warn_if_hostname_conflict, m_warn_if_unknown_ip,
+                        m_get_host_ips, m_check_system, m_os, m_sys_exit):
+        """
+        Test node_start function in calicoctl node command
+        """
+        # Set up mock objects
+        m_os.path.exists.return_value = False
+        m_get_host_ips.return_value = []
+        m_client.get_ip_pools.return_value = []
+        err = APIError("Test error message", Response())
+        m_docker_client.remove_container.side_effect = err
+
+        # Set up arguments
+        node_image = 'node_image'
+        log_dir = './log_dir'
+        ip = ''
+        ip6 = 'aa:bb::zz'
+        as_num = ''
+        detach = False
+        kubernetes = True
+
+        # Testing expecting APIError exception
+        self.assertRaises(APIError, node.node_start,
+                          node_image, log_dir, ip, ip6, as_num, detach, kubernetes)
```

Comment:
You should assert that `client.add_ip_pool` has a couple of calls here also since `get_ip_pools` returns and empty array.